multi tenancy blog (#224)
MechanicalRock/mechanicalrock.github.io: _posts/2022-05-23-fine-grained-access-control-for-multitenancy.markdown

---
layout: post
title: Fine Grained Access Control for Multi-tenancy
description: A how to guide for architecting multitenant backend applications
date: 2022-05-23
author: Leon Ticharwa
image: img/blog/multitenancy/multitenancy.png
tags: ['multitenant', 'typescript', 'tutorial', 'dynamodb', 'lambda', 'api gateway', 'authorization', 'authentication', 'apiGateway', 'jwt', 'json web token', 'cognito']
---

## Introduction

Architecting multi-tenant cloud native software can be quite challenging due to the broad scope of technical know-how required to execute such an undertaking.

> What's even more challenging is finding cohesive guides/tutorials that take a top down approach to the somewhat elusive topic of multitenancy. This article is an attempt to produce a distilled "how-to" guide for those who may find themselves bogged down while building multi-tenant applications. You clone the repo containing the reference code from here [reference code here](https://github.com/MechanicalRock/Multitenancy-AuthorizationAuthentication).

Reviewer: ```suggestion
What's even more challenging is finding cohesive guides/tutorials that take a top down approach to the somewhat elusive topic of multitenancy. This article is an attempt to produce a distilled "how-to" guide for those who may find themselves bogged down while building multi-tenant applications. You can clone the repo containing the reference code from here [reference code here](https://github.com/MechanicalRock/Multitenancy-AuthorizationAuthentication).
```

> What's even more challenging is finding cohesive guides/tutorials that take a top down approach to the somewhat elusive topic of multitenancy. This article is an attempt to produce a distilled "how-to" guide for those who may find themselves bogged down while building multi-tenant applications. You clone the repo containing the reference code from here [reference code here](https://github.com/MechanicalRock/Multitenancy-AuthorizationAuthentication).

Reviewer 2: Minor; consistency with multi-tenancy. Sometimes it's multi-tenancy, other times it's multitenancy (no dash).

> ## Some need to knows (Authorization and Authentication)

Reviewer 2: Suggestion; it might be worth introducing multi-tenancy. i.e. "What is multi-tenancy".

> For new developers, it is not immediately obvious that there is a difference between the terms Authorization and Authentication. Although related, these terms actually refer to two different concepts. Authentication verifies a user's identity whereas authorization verifies what a user is allowed to access once they have been authenticated. In short, authentication is about who is allowed in and authorization is about what they are allowed to access once they are in.

Reviewer 2: Minor; authorisation in Australian English (authorization is American English).

Reviewer 3: As per style guide: https://docs.google.com/document/d/1W1e0nKku0IYZ5spUXIeI8TG04sAELGvT67CHGHGwJT0/edit#heading=h.6cibo0m7iwd0

In this write up I'll demonstrate how one might go about using Cognito to develop the authentication and authorization layer in a contextualised scenario.

### JSON Web Tokens

JWT, JSON Web Token, is an open standard that is widely used to securely share authentication information (claims) between client and server. The standard is defined in the RFC7519 specification developed by the Internet Engineering Taskforce (IETF).

Valid JWTs contains three sections that are encoded as base64url strings separated by dot characters as shown below

```

1. Header

> 2. Payload

Reviewer 2: Suggestion; explain what the sections are used for.

#### Lambda Context Objects

In order to make use of the idea of a `tenantID` it is important to realise that there needs to be some way of retrieving the correct `tenantID` parameter and propagating it to the running lambda function's execution scope. This is where the idea of Lambda context objects comes into play.

Context objects can be modified to include custom parameters that can then be accessed during run time. With this in mind, it should be relatively straight forward to use the application's Lambda Authorizer to forward the retrieved `tenantID` as a custom parameter within the JSON output of the lambda authorizer.

When downstream lambdas are invoked they cam access the context object as a key within the event object. Take for example the following sample lambda event object.

```

> 2022-05-20T08:09:58.116Z	1c1c55e3-6ec1-47fb-962f-d94387e10480	INFO	{

Reviewer 2: Suggestion; It might be worth pointing out to the reader that it's not generally a good idea to print sensitive information out to logs.

  resourceId: 'hdo0nl',
  authorizer: {
    firstName: 'Anakin',
    lastName: 'Skywalker',
    org: 'galactic empire',
    tenantId: 'sith-inc-100',
    principalId: '24da3e19-0417-4e85-a9ed-078b9dcfb919',
    integrationLatency: 0
  },
  resourcePath: '/cart/{itemId}',
  httpMethod: 'PATCH',
  extendedRequestId: 'SajxWGotSwMFoIw=',
  requestTime: '20/May/2022:08:09:57 +0000',
  path: '/multiTenantStack/cart/1653034194717',

>   accountId: '935741529896',

Reviewer 2: Question; are these values sanitised? (Not real account IDs etc?)

  protocol: 'HTTP/1.1',
  stage: 'multiTenantStack',
  domainPrefix: 'htieqffa0l',
  requestTimeEpoch: 1653034197210,
  requestId: 'f260bfad-f7f3-4641-b492-6bbbff99de74',
  identity: {
    cognitoIdentityPoolId: null,
    accountId: null,
    cognitoIdentityId: null,
    caller: null,

>     sourceIp: '180.150.83.218',

Reviewer 2: Is this a real ip?

In essence, the `dynamodb:LeadingKeys` condition key is a mechanism for row level access control within DynamodDB tables. Employing the use of this mechanism guarantees a reliable and secure multitenant environment as well as the added benefit of a simplified data persistence layer.

![image](dynamdbFineGrainedAccess.png)

#### Lambda Authorizer Resource Policies

###### Token Verification failed

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Deny",
            "Action": "*",
            "Resource": "*"
        },
    ]
}
```

###### Token Verification passed

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": "execute-api:Invoke",
            "Resource": executeApiArn,
        },
        {

          Effect: 'Allow',
          Action: ['dynamodb:UpdateItem', 'dynamodb:PutItem', 'dynamodb:DeleteItem', 'dynamodb:Query'],
          Resource: this.cartTable,
          Condition: {
            'ForAllValues:StringLike': {
              'dynamodb:LeadingKeys': tenantId,
            },
          },
        },
    ]
}

> ```

Reviewer 2: Suggestion; conventionally we finish posts off with a summary. Re-cap what you covered and achieved in the tutorial, and what next steps might be. Point them to our contact page if they need consulting work.

---
layout: post
title: Fine Grained Access Control for Multi-tenancy
description: A how to guide for architecting multitenant backend applications
date: 2022-05-23
author: Leon Ticharwa
image: img/blog/multitenancy/multitenant.png
tags: ['multitenant', 'typescript', 'tutorial', 'dynamodb', 'lambda', 'api gateway', 'authorization', 'authentication', 'apiGateway', 'jwt', 'json web token', 'cognito']
---

## What is Multi-tenancy

Multi-tenancy is a software architecture that allows multiple users to access resources that reside in a shared computing environment. This architecture grants groups of users or individual users access to resources as defined by access boundaries put in place via the use of resource policies, IAM policies or some other similar mechanism. This concept is analogous to an apartment building where each tenant is given keys that only unlock their apartment door and other rooms within the apartment.

Architecting multi-tenant cloud native software can be quite challenging due to the broad scope of technical know-how required to execute such an undertaking.

> What's even more challenging is finding cohesive guides/tutorials that take a top down approach to the somewhat elusive topic of multitenancy. This article is an attempt to produce a distilled "how-to" guide for those who may find themselves bogged down while building multi-tenant applications. You can clone the repo containing the reference code from here [reference code here](https://github.com/MechanicalRock/Multitenancy-AuthorizationAuthentication).

Reviewer 3: You're burying the lede.

https://docs.google.com/document/d/1W1e0nKku0IYZ5spUXIeI8TG04sAELGvT67CHGHGwJT0/edit#heading=h.ihv636

Use your first paragraph is to hook the reader to get them to read on.

Introducing Multi-Tenancy can be under the Need to Knows below.

> ## Some need to knows (Authorization and Authentication)

Reviewer 3: ```suggestion
## Some Need to Knows
```

style: use title case

`Authentication and Authorisation` should be a sub-heading if you move Multi-Tenancy under here

General conventions I've seen use `authentication and authorisation` rather than `authorization and authentication`

I presume this stems from the fact that you need to generally need to authenticate an identity before making any authorisation decisions.

> ## Some need to knows (Authorization and Authentication)

Reviewer 3: The introductory information here is very useful, but is it detracting from the goal of this post - how to build a multi-tenant application?

Is it worth splitting this into two posts:
* An introductory post covering the various aspects of authentication and authorisation
* This post on multi-tenancy 
  * Refer to the terms you expect the reader to already be familiar with.  And link them to your other post for the detail.

For new developers, it is not immediately obvious that there is a difference between the terms Authorization and Authentication. Although related, these terms actually refer to two different concepts. Authentication verifies a user's identity whereas authorization verifies what a user is allowed to access once they have been authenticated. In short, authentication is about who is allowed in and authorization is about what they are allowed to access once they are in.

In this write up I'll demonstrate how one might go about using Cognito to develop the authentication and authorization layer in a contextualised scenario.

### JSON Web Tokens

> JWT, JSON Web Token, is an open standard that is widely used to securely share authentication information (claims) between client and server. The standard is defined in the RFC7519 specification developed by the Internet Engineering Taskforce (IETF).

Reviewer 3: Can't make as a suggestion for some reason...

JSON Web Token (JWT)

Abbreviations under style guide

}
```

- The principalId is the user id associated with the token sent by the client.
- If the API uses a usage plan and the apiKeySource is set to AUTHORIZER, the lambda authorizer output must include the usage plan's API keys as the `usageIdentifierKey` property value- The principalId is the user id associated with the token sent by the client.
- If the API uses a usage plan and the apiKeySource is set to AUTHORIZER, the lambda authorizer output must include the usage plan's API keys as the `usageIdentifierKey` property value

### Verifying tokens

Token verification is done in 3 steps.

1. Verify structure of token
2. Verify signature
3. Verify the claims

#### Verify structure of token

Confirm that the token contains three dot separated base64url strings. If the token does not conform to this structure then it is invalid.
The first string is a header string followed by a payload string and then finally the signature string as shown below.

```
<Header>.<Payload>.<Signature>
```

#### Verify signature

##### 1. Decode token

    To validate the JWT signature, the token must first be decoded.

##### 2. Compare local key ID (kid) to public key ID

> &nbsp;&nbsp;&nbsp;&nbsp;

Reviewer 3: are these required?

> i) Download and store your JWT's corresponding `JWK` (JSON Web Key) using the following url

Reviewer 3: any reason you have switched numbering style here?

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
`https://cognito-idp.{region}.amazonaws.com/{userPoolId}/.well-known/jwks.json`

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;

> substitute region and userPoolId with your user pool's region and user pool ID respectively

Reviewer 3: `Substitute`

Typo - new sentence

                "n": "1234567890",
                "use": "sig"
            }, {
                "kid": "5678example=",
                "alg": "RS256",
                "kty": "RSA",
                "e": "AQAB",
                "n": "987654321",
                "use": "sig"
            }]
        }
```

##### 3. Compare signature of the issuer to the signature of the tokens

The signature of the issuer is derived from the JWK with a kid that matches the kid of the JWT.
Each `JWK` contains an `n` parameter that contains the `modulus value` of the `RSA public key`.
This is the value that'll be used to derive the `issuer's signature`.
The JWK will need to be `converted to PEM format` before that can happen.

#### Verify the claims

1. Verify that the token is not expired.

2. The aud claim in an ID token and the client_id claim in an access token should match the app client ID that was created in the Amazon Cognito user pool.

3. The issuer (iss) claim should match your user pool. For example, a user pool created in the us-east-1 Region will have the following iss value:`https://cognito-idp.us-east-1.amazonaws.com/<userpoolID>`

4. Check the `token_use` claim.

> &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; If you are only accepting the access token in your web API operations, its value must be access.

Reviewer 3: again - any reason for the whitespace?

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; If you are only using the ID token, its value must be id.

> ## Scenario: Multi-tenant purchase tracking microservice

Reviewer 3: Burying the lede again.

As a reader I got bored and went elsewhere before I got to here.

If you don't want to split the article, suggest moving this to the top, as the introduction before signposting the need to introduce a bunch of concepts first.

Also add an `<a>` tag to allow the reader to skip introducing the terms if they are already familiar.

Suggest moving this introduction up to the top

![image](img/blog/multitenancy/arch.png)

1. Users authenticate with a username and password, the web app passes these to amazon cognito for validation.
2. If the supplied credentials (username and password) are valid, cognito creates a session and subsequently issues three (3) JWTs (JSON Web Tokens). The aforementioned tokens are id token, access token and a refresh token. The authenticated user can now send requests to api gateway along with with the id token in the headers section.
3. API gateway sends the received id token to a lambda function called an authorizer.
4. The authorizer function verifies the claims attached to the id token.
5. The authorizer returns a policy and context.
6. API Gateway evaluates the policy and forwards the request to a lambda function along with the authorizer generated context.

> 7. The lambda function writes/reads data according to the tenantId listed in the forwarded context.

Reviewer 3: number 7 is missing from your diagram.

8. A response is returned by the lambda function.

### Tenant Isolation

#### Tenant ID

> To create a secure `multi-tenant` environment there needs to be some notion of `tenant resource isolation`. In essence, `tenant A` should not be able to access the resources of `tenant B` and vice versa. To solve this problem it was decided that it is necessary to assign a unique identifier called `tenantID` to each user/customer. The aforementioned `tenantID` is a composite string that is generated and then attached to the calling user as a `custom attribute` during the registration stage.

Reviewer 3: Active Voice - style guide

```suggestion
To create a secure `multi-tenant` environment we require `tenant resource isolation`. In essence, `tenant A` should not be able to access the resources of `tenant B` and vice versa. We can achieve resource isolation assigning a unique identifier, `tenantID`, to each user/customer. The `tenantID` is a composite string that is generated and then attached to the calling user as a `custom attribute` during registration.
```

> After successful authentication the `tenantID` custom attribute becomes available as a parameter within the `ID token` as a parameter with the following key `custom:tenantID`. It is important to note that access tokens do not carry any of the user's custom attributes, only id tokens have this capability. In saying so, the application will exclusively use id tokens for authorization/authentication purposes.

Reviewer 3: ```suggestion
After successful authentication the `tenantID` custom attribute becomes available as a parameter within the `ID token` as a parameter with the following key `custom:tenantID`. It is important to note that access tokens do not carry any of the user's custom attributes, only id tokens have this capability. This application only uses id tokens for authentication/authorisation purposes.
```
